Products/CMFPlomino/PlominoReplicationManager.py

```diff
@@ -1304,6 +1329,82 @@ def saveFile(path, content):
         fileobj.write(content)
         fileobj.close()
 
+    security.declareProtected(READ_PERMISSION, 'exportDocumentsAsCSV')
+    def exportDocumentsAsCSV(self, docids=None):
+        logger.info("Starting documents CSV export...")
+        forms = self.getForms()
+        max_columns = 1000  # Arbitrary limit of 100000 columns. We'll remove excess columns later.
+        fieldnames = range(max_columns)
+        column_number_field_name_mapping = {0: "doc_id"}
+
+        for form in forms:
+            for field in form.getFormFields():
+                if field.id not in column_number_field_name_mapping.itervalues():
+                    column_number_field_name_mapping[len(column_number_field_name_mapping)] = field.id
+
+        if docids:
+            doc_ids = [self.getDocument(i) for i in docids]
+        else:
+            doc_ids = self.documents.keys()
+
+        number_of_docs = len(doc_ids)
+        logger.info("Exporting %s documents..." % number_of_docs)
+
+        # Returns a tuple of ({doc_data}, [new_field_names]) 
+        def _iterate_documents(doc_ids_list):
+            for i, doc_id in enumerate(doc_ids_list):
+                if i % 2000 == 0:
+                    transaction.abort()
+                yield self.documents[doc_id]
+
+        transaction.abort()
+        logger.info("All fieldnames checked. Starting document writing")
+
+        zope_export_folder_path = getattr(getConfiguration(), 'clienthome', "")
+        export_folder_path = os.path.join(zope_export_folder_path, 'export', self.id)
+        export_path = os.path.join(export_folder_path, '%s.csv' % self.id)
+        if os.path.isdir(export_folder_path):
+            # remove previous export
+            if os.path.exists(export_path):
+                os.remove(export_path)
+        else:
+            os.makedirs(export_folder_path)
+
+        with TemporaryFile(mode="w+b") as csv_data_tempfile:
+            writer = csv.DictWriter(csv_data_tempfile, fieldnames=fieldnames)
+
+            transaction.abort() 
+
+            for i, doc in enumerate(_iterate_documents(doc_ids)):
+                for field_name in doc.getItems():
+                    if field_name not in column_number_field_name_mapping.itervalues():
```

Comment:
how does checking for an id in an iterator work? don't you want to have map for fieldname -> col? That would seem more efficient.